### DISCUSSION TITLE
No More Footguns: Implement a "HW Configuration Guide" for Given OS, required Graphics API, and Game/SW

### DISCUSSION THREAD

**Discussion by the-burrito-triangle** [PCBox/PCBox#127] [Ideas]
The breadth of  emulated x86 hardware that PCBox (and 86Box) supports is quite impressive. The sheer number of overlapping options makes configuring a working--and more importantly--performant VM daunting. This is doubly true for those who didn't grow up with such HW and don't understand why one should choose such and such over other options for a given guest OS and game/SW that will be used with it. This overload ultimately leads to a "HW configuration hell" for the end user of PCBox.

Limiting HW choices for a particular OS would help immensely. And further pruning of CPUs, GPUs, and sound cards to only those that are known to actually perform well would be even better. I'm not saying to remove any functionality, but provide a "HW selection guide" to help configure the system based on desired outcome: e.g., playing game X with graphics API Y on Windows version Z.

As a concrete example, let's say we want a Windows 98SE guest. Many games are 2D DDraw based and don't need a powerful GPU; for example, Diablo I. Then pretty much any S3 graphics card (most of which only do 2D) would be fine. But there are A LOT of S3 cards available, so which one works best? The guide should pick the most tested and performant option for the user. If Glide or D3D were needed, another well-tested GPU should be suggested instead. As for the CPU, I'd only focus on what games were actually targeting: Intel parts, particularly the Pentium I to III. If we were wanting to use Windows 3.1 then an older, pre Pentium CPU might be more desirable.

Another example would be Windows NT, for which some HW options do not have drivers: so don't show these options in the HW list, etc. Additionally, some OSes (like ReactOS) require specific CPU instructions (SSE), so only certain CPUs would be a valid option, like the PIII coppermine and newer.

Ultimately, only a few HW options are really needed (maybe 4-5 CPUs, 3 GPUs and 2-3 sound cards). And at that point, optimizations can actually be done since we are targeting a much smaller number of configurations specific for a given OS. And the 2D S3 cards could possibly use DOSBox-X optimizations to squeeze out better performance for DDraw games. Similarly, the older Voodoo options could possibly use OpenGlide to pass OpenGL commands to the Host for better performance. Trying to emulate everything under the sun is nice for testing and special situations, but it doesn't make for a simple, user friendly setup experience due to myriad options that seemingly don't matter in the end: e.g., multiple vendors for the same graphics card--just keep the one that works best rather than provide two options that look redundant. In some cases, a catalog of known Windows bugs for a given HW / driver version might be used to help select different HW--even if it hypothetically should have been fine to use.

Anyways, those are my initial thoughts from having played around with PCBox and 86Box recently. I'm impressed that I got ReactOS to run on PCBox with a PIII setup. While I like the incredible scope of PCBox and 86Box, a slimmed down version would be immensely more useful for retro gaming. Especially if optimizations to reduce CPU load and increase performance could be made for select Windows guest configurations.

**Comment by OBattler** [comment]
Hello, 86Box lead developer here.

> Limiting HW choices for a particular OS would help immensely.

There's so many OS'es and versions of those OS'es, all with different requirements and lists of supported hardware that that wouldn't be practical at all.

> but provide a "HW selection guide" to help configure the system based on desired outcome: e.g., playing game X with graphics API Y on Windows version Z.

But we do have something similar in the eventual plans - a wizard where you can choose from among the most popular OS'es and the emulator then sets up the recommended hardware for it.

> As for the CPU, I'd only focus on what games were actually targeting: Intel parts, particularly the Pentium I to III.

Games were quite deliberately targeting AMD parts as well, just usually not Cyrix or WinChip.

> Additionally, some OSes (like ReactOS) require specific CPU instructions (SSE), so only certain CPUs would be a valid option, like the PIII coppermine and newer.

Since when does ReactOS require SSE? I recall it did not, and they even used 86Box to test their ISA PnP implementation. In fact, I even recall it working on the emulated Pentium, so not even i686 may be required.

> And at that point, optimizations can actually be done since we are targeting a much smaller number of configurations specific for a given OS. And the 2D S3 cards could possibly use DOSBox-X optimizations to squeeze out better performance for DDraw games.

Those are already optimized as much they can be - their 2D accelerators are even threaded.

> Similarly, the older Voodoo options could possibly use OpenGlide to pass OpenGL commands to the Host for better performance.

That would defeat the point of emulating them to provide a period gaming experience.

> e.g., multiple vendors for the same graphics card--just keep the one that works best rather than provide two options that look redundant.

We're planning to eventually clean that up by eventually moving vendor selection to device configuration and defaulting it to the best one, because even I don't like how cluttered the list has become.

The reason it's been taking us so long to carry out these plans is because we're a small team. Just adding MDS+MDF support (and doing all the CD-ROM fixes and clean-ups required first) has taken me longer than planned due to alsohaving to fix the issues reported on GitHub and Discord, let alone doing stuff like the wizard or the decluttering. But we're aware and will eventually get there.

**Reply by bachimusprime** [reply]
OBattler here is absolutely spot on. Yeah, most of these suggestions are a bit of a no-go, sorry.
